add typedef and namespace

ygdrasil-io opened the pull request.

ygdrasil-io committed:
Add TypeDef parsing support

Introduced support for parsing `typedef` elements in WebIDL. Implemented a `TypeDefParser` class, updated `WebIdlParserType` to include `TypeDef`, and added a new `IdlTypeDef` model for improved functionality.

ygdrasil-io committed:
Refactor `parseChildren` to remove redundant parameter.

Replaced `parseChildren(termToken: String?)` with a simplified overload in `ElementParser` and updated its usage in `TypeDefParser`. This improves code clarity and avoids unnecessary parameter passing where not required.

ygdrasil-io committed:
Add support for typedef parsing in WebIDL model

Introduce typedef parsing functionality to the WebIDL parser, including updates to the `TypeDefParser` and `IdlModel` classes to handle typedef definitions. Added corresponding test cases to verify the parsing logic, ensuring typedefs are correctly modeled and included in the overall structure.

ygdrasil-io committed:
Add namespace and constant parsing support

Introduces parsing and model support for namespaces and their constants in the WebIDL parser. Updates include adding `NamespaceParser`, `ConstantParser`, and relevant model adjustments to handle namespaces and constant definitions. Tests and IDL resources are updated to validate the new functionality.

ygdrasil-io committed:
Refactor IdlConstant: remove unused parentDictionary logic

Removed the `parentDictionary` property and associated logic, which were unused in the `IdlConstant` class. Additionally, changed `defaultValue` initialization to `lateinit` to streamline builder usage.

ygdrasil-io committed:
Remove unused 'sourcePackage' property from IdlNamespace.

The 'sourcePackage' property and its related code were redundant and served no functional purpose. This cleanup improves code clarity and removes unnecessary complexity in the NamespaceParser and IdlNamespace classes.

ygdrasil-io committed:
Remove unused finishModel function in IdlNamespace

The `finishModel` function was removed as it was not being used anywhere in the code. This cleanup simplifies the class and reduces unnecessary code, improving maintainability.

ygdrasil-io committed:
Rename 'constantes' to 'constants' for consistency

Updated variable and method names from 'constantes' to 'constants' in both `ParserTest` and `IdlNamespace` classes. This ensures alignment with standard naming conventions and improves code readability.

ygdrasil-io committed:
Remove unused typedef ATypeDef3 from test.idl

ATypeDef3 was commented out and left unused in the file. Removing it simplifies the code and avoids unnecessary clutter in the test resource.

fabmax commented on webidl-util/src/main/kotlin/de/fabmax/webidl/parser/ConstantParser.kt:
```diff
@@ -0,0 +1,24 @@
+    override suspend fun parse(): String {
+        popToken("const")
+        val type = parseType()
+        val tokens = popUntilPattern(";") ?: parserException("Failed parsing member")
```
I assume the exception message should be "Failed parsing constant"

ygdrasil-io replied:
Fixed

ygdrasil-io committed:
Update error message in ConstantParser

Adjusted the error message for parsing failure from "member" to "constant" to improve clarity and accuracy. This ensures the message aligns with the context of parsing constants.
